notion-server/src/ethereum/mod.rs

```diff
@@ -152,45 +169,84 @@ impl Simulation {
     */
 
     pub fn get_shard_block(&self, args: args::GetShardBlock) -> Result<args::ShardBlock> {
-        if let Some(shard_chain) = self.shard_chains.get(args.shard_chain_index as usize) {
-            if let Some(shard_block) = shard_chain
-                .shard_blocks
-                .get(args.shard_block_index as usize)
-            {
-                Ok(args::ShardBlock::from(shard_block))
-            } else {
-                Err(Error::OutOfBounds {
-                    message: format!(
-                        "the shard chain at index '{}' does not contain a block at index '{}'",
-                        args.shard_chain_index, args.shard_block_index
-                    ),
-                })
-            }
-        } else {
-            Err(Error::OutOfBounds {
-                message: format!(
-                    "no shard chain exists at index '{}'",
-                    args.shard_chain_index
-                ),
-            })
-        }
+        let shard_chain = self
+            .shard_chains
+            .get(args.shard_chain_index as usize)
+            .context(OutOfBounds {
+                what: WhatBound::ShardChain,
+                index: args.shard_chain_index as usize,
+            })?;
+
+        let shard_block = shard_chain
+            .shard_blocks
+            .get(args.shard_block_index as usize)
+            .context(OutOfBounds {
+                index: args.shard_block_index as usize,
+                what: WhatBound::ShardBlock(args.shard_chain_index),
+            })?;
+
+        Ok(args::ShardBlock::from(shard_block))
     }
 }
 
-pub mod args {
+mod b64_vec {
+    use serde::de::{Deserialize, Deserializer, Error as _, Unexpected};
+    use serde::Serializer;
+
+    pub fn serialize<T, S>(bytes: T, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        T: AsRef<[u8]>,
+        S: Serializer,
+    {
+        let txt = base64::encode(bytes.as_ref());
+        serializer.serialize_str(&txt)
```

Comment:
Not sure I follow why this is necessary -- if we already have it as a string, what need to serialize further?

Reply:
Custom (de)serialization with serde is probably one of the hardest things to learn. If you'd like to go over this on zoom, we can!

The summary is that the `base64::encode` creates a string, then `serialize_str` is telling the JSON encoder that the next thing to write is a string.

Reply:
Hmmm sounds like I'll set aside some time to do this on my own.  FWIW, I wonder whether this complexity is necessary.  In my initial brief investigations into Serde/Rocket, auto serialization/deserialization was very straightforward, so it seems odd that this extra stuff is required -- is there a reason the default de/serialize to/from structs isn't powerful enough?

Reply:
The default behaviour for `Vec<u8>` serialization/deserialization is to use a JSON array full of `u8` values.

Since it seems like we want to use base64 (or hex), we need to override the defaults. The actual implementation here isn't that complex. All it does is read a string value and convert it to a byte array using base64. The rest is just serde boilerplate.